File: tools/ghidra/ExportPS1Functions.py
# ExportPS1Functions.py — Ghidra Script
#
# Exports all non-thunk functions from a PS1 binary loaded in Ghidra to a CSV
# file that ps1xAnalyzer can consume to skip its own function-boundary detection.
#
# Usage (Ghidra Script Manager):
#   1. Load your PS1 EXE / BIN in Ghidra and run auto-analysis.
#   2. Open Script Manager → Run Script → select this file.
#   3. When prompted, choose the output path (default: <project>/ps1_functions.csv).
#
# Alternatively run headless:
#   analyzeHeadless /path/to/project ProjectName \
#       -import SLUS_000.05 \
#       -postScript ExportPS1Functions.py /tmp/ps1_functions.csv \
#       -scriptPath /path/to/tools/ghidra
#
# CSV format (no header):
#   Name,StartAddress,EndAddress,Size
#   func_8012D030,0x8012D030,0x8012D09C,0x6C
#
# Adapted from PS2Recomp/ps2xRecomp/tools/ghidra/ExportPS2Functions.py

import os
from ghidra.program.model.symbol import SymbolType

# ─── Configuration ────────────────────────────────────────────────────────────
DEFAULT_OUTPUT = os.path.join(
    currentProgram.getDomainFile().getParent().getProjectLocator().getProjectDir().toString(),
    "ps1_functions.csv",
)
SKIP_THUNKS   = True   # Exclude thunk functions (wrappers with a single JR)
SKIP_EXTERNAL = True   # Exclude external / imported symbols
MIN_SIZE      = 4      # Skip functions smaller than N bytes (likely false positives)

# ─── Ask user for output path ─────────────────────────────────────────────────
output_path = askString(
    "Export PS1 Functions",
    "Output CSV path:",
    DEFAULT_OUTPUT,
)

# ─── Iterate functions ────────────────────────────────────────────────────────
function_manager = currentProgram.getFunctionManager()
listing = currentProgram.getListing()
addr_factory = currentProgram.getAddressFactory()

rows = []
for func in function_manager.getFunctions(True):   # True = forward iteration
    if SKIP_THUNKS and func.isThunk():
        continue
    if SKIP_EXTERNAL and func.isExternal():
        continue

    entry = func.getEntryPoint()
    body  = func.getBody()
    size  = body.getNumAddresses()

    if size < MIN_SIZE:
        continue

    # Compute end address (inclusive last byte of last range)
    end_addr = body.getMaxAddress()

    name       = func.getName()
    start_hex  = "0x{:08X}".format(entry.getOffset())
    end_hex    = "0x{:08X}".format(end_addr.getOffset())
    size_hex   = "0x{:X}".format(size)

    rows.append("{},{},{},{}".format(name, start_hex, end_hex, size_hex))

# ─── Write output ─────────────────────────────────────────────────────────────
with open(output_path, "w") as f:
    f.write("Name,StartAddress,EndAddress,Size\n")
    for row in rows:
        f.write(row + "\n")

print("[ExportPS1Functions] Exported {} functions to: {}".format(len(rows), output_path))

File: tools/ghidra/ExportPS1Functions.java
// ExportPS1Functions.java — Ghidra Script
//
// Java equivalent of ExportPS1Functions.py.
// Preferred when running via Ghidra's auto-analysis (headless mode) since
// Java scripts compile ahead of time and run faster than Jython.
//
// Usage:
//   analyzeHeadless /path/to/project ProjectName \
//       -import SLUS_000.05 \
//       -postScript ExportPS1Functions.java /tmp/ps1_functions.csv \
//       -scriptPath /path/to/tools/ghidra
//
// @category PS1Recomp
// @author ps1-recomp project

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.address.AddressSetView;

import java.io.FileWriter;
import java.io.PrintWriter;

public class ExportPS1Functions extends GhidraScript {

    private static final int  MIN_SIZE     = 4;
    private static final boolean SKIP_THUNKS   = true;
    private static final boolean SKIP_EXTERNAL = true;

    @Override
    public void run() throws Exception {
        // Output path: first script argument, or default beside project dir
        String outputPath = getScriptArgs().length > 0
            ? getScriptArgs()[0]
            : currentProgram.getDomainFile()
                .getParent()
                .getProjectLocator()
                .getProjectDir()
                .getAbsolutePath() + "/ps1_functions.csv";

        FunctionManager fm = currentProgram.getFunctionManager();
        int count = 0;

        try (PrintWriter writer = new PrintWriter(new FileWriter(outputPath))) {
            writer.println("Name,StartAddress,EndAddress,Size");

            for (Function func : fm.getFunctions(true)) {
                if (SKIP_THUNKS   && func.isThunk())    continue;
                if (SKIP_EXTERNAL && func.isExternal())  continue;

                AddressSetView body = func.getBody();
                long size = body.getNumAddresses();

                if (size < MIN_SIZE) continue;

                long startOff = func.getEntryPoint().getOffset();
                long endOff   = body.getMaxAddress().getOffset();

                writer.printf("%s,0x%08X,0x%08X,0x%X%n",
                    func.getName(), startOff, endOff, size);
                count++;
            }
        }

        printf("[ExportPS1Functions] Exported %d functions to: %s%n", count, outputPath);
    }
}
